pkg/Cpanel/Security/Advisor/Assessors/Trojans.pm

```diff
@@ -73,7 +91,611 @@ sub _check_for_libkeyutils {
+        ( $host, $port ) = ( $1, $2 );
+    }
+
+    return if !$host;
+    return if !$port;
+
+    my $sock = IO::Socket::INET->new(
+        PeerAddr => $host,
+        PeerPort => $port,
+        Proto    => 'tcp',
+        Timeout  => 5,
+    ) or return;
+
+    $ssh_banner = readline $sock;
+    close $sock;
+    return if !$ssh_banner;
+    chomp $ssh_banner;
+
+    if ( $ssh_banner =~ m{ \A SSH-2\.0-[0-9a-f]{22,46} }xms ) {
+        $self->add_bad_advice(
+            'text'       => [ "Trojan sshd binary check: suspicious checksums/hashes were found that could indicate the existance of the Ebury rootkit." ],
+            'suggestion' => [
+                'Check the following for more information "[output,url,_1,We Live Security More Information,_2,_3]"',
+                'http://www.welivesecurity.com/2014/02/21/an-in-depth-analysis-of-linuxebury/',
+                'target',
+                '_blank'
+            ],
+        );
+    }
+}
+
+sub _check_for_ebury_ssh_shmem {
+    my ($self) = @_;
+    # As far as we know, sshd sholudn't be using shared memory at all, so any usage is a strong 
+    # sign of ebury.
+    return if !defined( $IPCS_REF->{root}{mp} );
+    my $PROCESS_REF;
+    for my $href ( @{ $IPCS_REF->{root}{mp} } ) {
+        my $shmid = $href->{shmid};
+        my $cpid  = $href->{cpid};
+        if ( $PROCESS_REF->{$cpid}{CMD} && $PROCESS_REF->{$cpid}{CMD} =~ m{ \A /usr/sbin/sshd \b }x ) {
+            $self->add_bad_advice(
+                'text'       => [ "sshd shared memory check: A shared memory segment created by sshd process exists." ],
+                'suggestion' => [
+                    'Check pages 67-68 from the following for more information "[output,url,_1,We Live Security More Information,_2,_3]"',
+                    'http://www.welivesecurity.com/wp-content/uploads/2014/03/operation_windigo.pdf',
+                    'target',
+                    '_blank'
+                ],
+            );
+        }
+    }
+}
+
+sub get_ipcs_hash {
+    my %hash;
+    my $header = 0;
+    # For now, all we need is shared memory segment owner and creator-pid, but the data 
+    # structure is extensible.
+    # ipcs -m -p
+    #
+    #------ Shared Memory Creator/Last-op --------
+    #shmid      owner      cpid       lpid
+    #2228224    root       992        992
+    #2588673    root       1309       1315
+    #2195458    root       985        985
+    #2621443    root       1309       1315
+    for ( split /\n/, timed_run( 0, 'ipcs', '-m', '-p' ) ) {
+        if ( $header == 0 ) {
+            $header = 1 if m/^ shmid \s+ owner \s+ cpid \s+ lpid \s* $/ix;
+            next;
+        }
+        my @ipcs = split( /\s+/, $_, 5 );
+        push @{ $hash{ $ipcs[1] }{mp} }, {    # Key by owner, type 'mp' (-m -p output)
+            'shmid' => $ipcs[0],
+            'cpid'  => $ipcs[2],
+            'lpid'  => $ipcs[3]
+        };
+    }
+    return \%hash;
+}
+
+sub timed_run { 
```

Comment:
FYI: This function assumes the alarm is currently not set, which is a fair assumption but worth noting.  See Cpanel::Alarm for a more complete alarm system.